### PR TITLE
test/new-e2e/trace-agent: add basic e2e test

### PR DESCRIPTION
**Pull request by knusbaum**
### What does this PR do?

This PR adds a new, simple e2e test for the trace agent, generating some traces and checking that intake receives what we expect.

### Motivation

This will serve as a basic check of the agent, as well as serving as a template for more complicated e2e tests.

### Describe how to test/QA your changes

N/A 

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: test/new-e2e/pkg/components/remotehost.go
Comment (knusbaum) [inline test/new-e2e/pkg/components/remotehost.go:170]: This might be useful for the future, even if we don't want it right now.
It allows us to create a TCP tunnel into the host and communicate with things on the remote.

Initially, I was going to use this to run a tracer locally and send traces to the remote agent, before discussions determined that we'd need a remote tracer anyway.

I won't put it into this PR, but it's here now to see if it's something I should move into another PR or just delete.
WDYT?

File: test/new-e2e/pkg/utils/infra/stack_manager.go
Comment (knusbaum) [inline test/new-e2e/pkg/utils/infra/stack_manager.go:244]: TODO: revert

File: test/new-e2e/pkg/components/remotehost.go
Comment (knusbaum) [inline test/new-e2e/pkg/components/remotehost.go:147]: I'm on the fence. I could just expose `h.client.Dial` as it is, so you can dial anything from the host, but I'm not sure if that's desirable.

File: test/new-e2e/tests/apm/docker_test.go
Comment (ahmed-mez) [inline test/new-e2e/tests/apm/docker_test.go:91]: While I agree these logs can be spammy, they can also be super useful for debugging. Note that they only appear if we run the tests with the `-v` (verbose) option. I'd vote for keeping them unless there other reasons why you deleted them.

**Review by ahmed-mez — COMMENTED** [review]
Looks good, just added some comments!

Before merging, let's make sure all the tests pass locally since they're skipped on the CI currently.

File: test/new-e2e/go.mod
Comment (knusbaum) [inline test/new-e2e/go.mod:10]: Required just for the moment. depends on https://github.com/DataDog/test-infra-definitions/pull/582

File: test/new-e2e/pkg/environments/aws/host/host.go
Comment (knusbaum) [inline test/new-e2e/pkg/environments/aws/host/host.go:175]: This seems to cause issues with a fresh VM, due to multiple things running apt at the same time. Still figuring out how to get around that:
```
         +  pulumi:pulumi:Stack REDACTED **creating failed (49s)** 1 error; 2 messages
        Diagnostics:
          pulumi:pulumi:Stack (REDACTED):
            fakeintake task private ip found: REDACTED
            Waiting for fakeintake at REDACTED to be healthy
            error: update failed
        
          command:remote:Command (remote-aws-vm-cmd-apt-install-docker.io-REDACTED):
            error: Process exited with status 100: running "DEBIAN_FRONTEND=\"noninteractive\" sudo apt-get install -y docker.io":
            E: Could not get lock /var/lib/dpkg/lock-frontend. It is held by process 2147 (apt-get)
            E: Unable to acquire the dpkg frontend lock (/var/lib/dpkg/lock-frontend), is another process using it?
```

File: test/new-e2e/tests/apm/docker_test.go
Comment (knusbaum) [reply test/new-e2e/tests/apm/docker_test.go:91]: I will add these back.

File: test/new-e2e/tests/apm/docker_test.go
Comment (knusbaum) [reply test/new-e2e/tests/apm/docker_test.go]: You're right, it's not required for the docker agent, but is for the agent on VM.

File: test/new-e2e/tests/apm/tests.go
Comment (ajgajg1134) [inline test/new-e2e/tests/apm/tests.go:21]: is this line required here?

File: test/new-e2e/tests/apm/tests.go
Comment (ajgajg1134) [inline test/new-e2e/tests/apm/tests.go:22]: nit: I think most of these are flipped? expected, actual

File: test/new-e2e/tests/apm/tracegen.go
Comment (ajgajg1134) [inline test/new-e2e/tests/apm/tracegen.go:50]: does this get used anywhere? I did a quick search on this page and didn't see it (but maybe github is being weird)

**Review by ajgajg1134 — CHANGES_REQUESTED** [review]
Overall looks good for our e2e tests, although I can't really comment too much on the changes to the components themselves (but I'm sure e2e team can look at those)

Just a few small change requests :)

File: test/new-e2e/tests/apm/vm_test.go
Comment (ahmed-mez) [inline test/new-e2e/tests/apm/vm_test.go:1]: Perhaps the e2e-testing reviewer can chime in, but I think calling `fixAgent` and `waitRemotePort` could be done only once in a new `VMFakeintakeSuite.SetupSuite` method, as opposed to calling them in every test-case.

See the `BaseSuite.SetupSuite` method [here](https://github.com/DataDog/datadog-agent/blob/82382e2cc2e4a53d55a25f6907afdd5c3f158ced/test/new-e2e/pkg/e2e/suite.go#L425-L431), we could override it.

**Review by ahmed-mez — COMMENTED** [review]
Nice refactoring of the existing tests 👏

File: test/new-e2e/tests/apm/vm_test.go
Comment (knusbaum) [reply test/new-e2e/tests/apm/vm_test.go:1]: Great idea. I didn't think `SetupSuite` had access to the infrastructure yet, but it looks like it does!

Comment (knusbaum) [reply test/new-e2e/tests/apm/vm_test.go:1]: Done.

**Review by ahmed-mez — APPROVED** [review]
This is great! 💯

(let's just make sure the tests succeed locally before merging since we're not running in the CI just yet - blocked by APL-2786)

Edit: the tests run in CI now.

**Review by ajgajg1134 — APPROVED** [review]

File: test/new-e2e/tests/apm/vm_test.go
Comment (vboulineau) [inline test/new-e2e/tests/apm/vm_test.go:174]: Why not `Eventually` for this one?

File: test/new-e2e/tests/apm/docker_test.go
Comment (ahmed-mez) [inline test/new-e2e/tests/apm/docker_test.go:40]: Added this because the default stack name caused a collision between test suites and errors in the CI

File: test/new-e2e/tests/apm/vm_test.go
Comment (knusbaum) [inline test/new-e2e/tests/apm/vm_test.go:176]: @vboulineau Not sure how to get the e2e.Context you mentioned. I don't see where I would get access to one anywhere, unless I need to hang on to a reference from the `Init` function.

Comment (pducolin) [reply test/new-e2e/tests/apm/vm_test.go:176]: You'd need to add `context.Context` to `e2e.Context` - currently `e2e.Context` defines only a `T()` function to return a `*testing.T` instance. 

The suite initialiser needs some changes too, as we need to create a context, probably at `suite.init()`. Then the function can be `RemoteHost.DialRemotePort(port)`.

We can help with the implementation, but it looks like a bigger change than expected, so I suggest keeping your implementation and refactor it in a follow up, or wait for us to implement the change, or again open a new PR that introduces only the `remoteHost.DialPort` function.

File: test/new-e2e/tests/apm/docker_test.go
Comment (pducolin) [inline test/new-e2e/tests/apm/docker_test.go:30]: **💬 suggestion**
No need to parse flags in tests, `go test` does it out of the box, except for `TestMain`

File: test/new-e2e/pkg/environments/aws/host/host.go
Comment (pducolin) [inline test/new-e2e/pkg/environments/aws/host/host.go:130]: **💭 thought**
Having docker in an host environment prevents from having the docker client on the output environment struct. I'd rather use the `awsdocker` environment that has a docker client. The provider there takes care of defaulting to AMIs that have docker already installed, which saves some setup time. But we need a way to allow installing the agent on docker or on the host, so maybe it makes sense to use `awshost` for this goal. 

One way to address the issue of extending default providers is to export the run function, so that you can call it from your custom provider with a custom env based on `awshost`. Would that work for you ?

File: test/new-e2e/tests/apm/vm_test.go
Comment (pducolin) [inline test/new-e2e/tests/apm/vm_test.go:44]: **💬 suggestion**
Not a blocker, you can use `embed` to import it from a yaml file and leverage formatting

File: test/new-e2e/tests/apm/docker_test.go
Comment (pducolin) [reply test/new-e2e/tests/apm/docker_test.go:34]: You can fix this also by having one base suite and than one implementation per `Test*`, stack name use the suite name as an ID

**Review by pducolin — APPROVED** [review]

**Review by zhuminyi — APPROVED** [review]

**Pull request merged**
